client/lib/query-manager/media/index.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * External dependencies
+ */
+import moment from 'moment';
+import { every, includes } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import PaginatedQueryManager from '../paginated';
+import MediaQueryKey from './key';
+import { DEFAULT_MEDIA_QUERY } from './constants';
+
+/**
+ * MediaQueryManager manages media which can be queried and change over time
+ */
+export default class MediaQueryManager extends PaginatedQueryManager {
+	/**
+	 * Returns true if the media item matches the given query, or false
+	 * otherwise.
+	 *
+	 * @param  {Object}  query Query object
+	 * @param  {Object}  media Item to consider
+	 * @return {Boolean}       Whether media item matches query
+	 */
+	matches( query, media ) {
+		return every( { ...DEFAULT_MEDIA_QUERY, ...query }, ( value, key ) => {
+			switch ( key ) {
+				case 'search':
+					if ( ! value ) {
+						return true;
+					}
+
+					return media.title && includes( media.title.toLowerCase(), value.toLowerCase() );
+
+				case 'mime_type':
+					if ( ! value ) {
+						return true;
+					}
+
+					// See: https://developer.wordpress.org/reference/functions/wp_post_mime_type_where/
+					return new RegExp( `^${
+						value
+							// Replace wildcard-only, wildcard group, and empty
+							// group with wildcard pattern matching
+							.replace( /(^%|(\/)%?)$/, '$2.+' )
+							// Split subgroup and group to filter
+							.split( '/' )
+							// Remove invalid characters
+							.map( ( group ) => group.replace( /[^-*.+a-zA-Z0-9]/g, '' ) )
+							// If no group, append wildcard match
+							.concat( '.+' )
+							// Take only subgroup and group
+							.slice( 0, 2 )
+							// Finally, merge back into string
+							.join( '/' )
+					}$` ).test( media.mime_type );
+
+				case 'post_ID':
+					// Note that documentation specifies that query value of 0
+					// will match only media not attached to posts, but since
+					// those media items assign post_ID of 0, we can still
+					// match by equality
+					return value === media.post_ID;
+
+				case 'after':
+				case 'before':
+					const queryDate = moment( value, moment.ISO_8601 );
+					const comparison = /after$/.test( key ) ? 'isAfter' : 'isBefore';
+					return queryDate.isValid() && moment( media.date )[ comparison ]( queryDate );
+			}
```

Comment:
>In addition, it's worth to consider whether we might want to query media by:

I should have clarified, but the intent of a QueryManager is to recreate the querying behavior from our REST API endpoints, in this case [the `GET /sites/%s/media` endpoint](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/media/) which does not include these parameters (maybe they should!).